### PR TITLE
Initial charm configuration

### PR DESCRIPTION
**Pull request by sudeephb**

### CODE REVIEW

File: src/config.yaml
Comment (mkalcok) [inline src/config.yaml:3]: Is the snap in stable channel already/

File: src/metadata.yaml
Comment (mkalcok) [inline src/metadata.yaml]: Please rewrite the description to be easier to read. Suggestion:
```
This charm install a Prometheus exporter that queries "apt" and "dpkg" to export following metrics about packages on the unit:
* Installed packages
* Broken packages
* Packages that can be upgraded
* Packages that can be auto-removed
In addition, a "pkg_reboot_required" metric is exported that indicates whether the system requires a reboot due to the new packages.
```

File: src/metadata.yaml
Comment (mkalcok) [inline src/metadata.yaml]: If memory serves me well, the `series` should be entirely removed since we have `charmcraft.yaml` now. I can't find more authoritative source right now than [this](https://bugs.launchpad.net/charm-elasticsearch/+bug/1983661).

File: src/metadata.yaml
Comment (mkalcok) [inline src/metadata.yaml:21]: Is there a reason to use `http` instead of [`scrape interface`](https://git.launchpad.net/charm-prometheus2/tree/src/metadata.yaml#n26) provided by `prometheus2` charm? iirc, the `scrape` should be preferred over `http`

File: src/reactive/prometheus-apt-exporter.py
Comment (mkalcok) [inline src/reactive/prometheus-apt-exporter.py]: forgotten comment?

File: src/reactive/prometheus-apt-exporter.py
Comment (mkalcok) [inline src/reactive/prometheus-apt-exporter.py]: Using [`hookenv.charm_dir()`](https://charm-helpers.readthedocs.io/en/latest/api/charmhelpers.core.hookenv.html?highlight=charm_dir#charmhelpers.core.hookenv.charm_dir) instead of `getcwd()` might be more appropriate.

File: src/reactive/prometheus-apt-exporter.py
Comment (mkalcok) [inline src/reactive/prometheus-apt-exporter.py]: I strongly believe that the port of the exporter should be configurable via charm. Especially since we expect this charm to be deployed on machines with other charms/services.

File: tests/integration/bundle.yaml.j2
Comment (mkalcok) [inline tests/integration/bundle.yaml.j2]: missing newline at the end of the file

File: tests/integration/conftest.py
Comment (mkalcok) [inline tests/integration/conftest.py]: missing newline at the end of the file

File: tests/integration/test_charm.py
Comment (mkalcok) [inline tests/integration/test_charm.py]: missing newline at the end of the file

File: src/reactive/prometheus-apt-exporter.py
Comment (mkalcok) [inline src/reactive/prometheus-apt-exporter.py:31]: Out of curiosity, is this flag necessary? Without the `juju-info` relation the unit is not deployed at all.

File: src/reactive/prometheus-apt-exporter.py
Comment (mkalcok) [inline src/reactive/prometheus-apt-exporter.py]: This is intended to refresh exporter snap, right? I have two suggestions here:

1. Just by removing the flags, you are not guaranteed that the `install_packages()` function will be triggered in "this" hook execution. It may very well happen that it will execute only the next time a hook is triggered (which might be on the next `update-status`). It'd be better to un-set flags and call `install_packages()` explicitly.

2. Scenario, in which user installs exporter snap from snap store initially but then decides to upload its own snap via resource, may not work. Looking at the [`install()`](https://git.launchpad.net/layer-snap/tree/lib/charms/layer/snap.py#n57) function from `layer.snap`. The first thing that it's doing is checking whether the flag for "snap already installed" exits. If it does, then it just triggers refresh. So in scenario when user adds his/her own snap via resource later, it'd just trigger refresh from snap store, disregarding local snap.

Feel free to mark this comment as "Resolved" if you don't care about this niche scenarios since re-work to Operator framework is likely in near future.

File: src/reactive/prometheus-apt-exporter.py
Comment (mkalcok) [inline src/reactive/prometheus-apt-exporter.py]: `set_state` and `remove_state` were deprecated in favor of `set_flag` and `clear_flag` (https://charmsreactive.readthedocs.io/en/latest/charms.reactive.flags.html?highlight=remove_state#charms.reactive.flags.remove_state)

File: src/reactive/prometheus-apt-exporter.py
Comment (mkalcok) [inline src/reactive/prometheus-apt-exporter.py]: same issue as above (and below) clearing the flags does not ensure immediate execution of decorated functions.

File: src/reactive/prometheus-apt-exporter.py
Comment (mkalcok) [inline src/reactive/prometheus-apt-exporter.py:129]: Setting unit status in reactive charms is tricky when it can come from different sources. For example in your case there are 3 functions that call `status_set` and you don't know in which order they'll be executed. For this reason, reactive charms ten to utilise flags and then final `atexit` decorated function that sets the final states after considering each possibility ([example](https://github.com/charmed-kubernetes/charm-kubernetes-control-plane/blob/ce8cea2b9bc6d8dfd60a0d3455be3aa97fdd8eb9/reactive/kubernetes_control_plane.py#L778))

**Review by mkalcok — CHANGES_REQUESTED** [review]
I left multiple comments, out of all of them, I'd consider only "configurable port number" and "metadata.yaml description" to be blockers. For the rest of them I'm going to leave it up to you, especially since rewrite to Operator framework is likely.

File: CONTRIBUTING.md
Comment (rgildein) [inline CONTRIBUTING.md]: Can we stick with `-e reformat`? It's handy if we use same names everywhere.

File: CONTRIBUTING.md
Comment (rgildein) [inline CONTRIBUTING.md]: `-e func`

File: src/files/grafana-dashboards/pkg.json
Comment (rgildein) [inline src/files/grafana-dashboards/pkg.json:4]: This should not be template? Like [this](https://opendev.org/openstack/charm-ceph-osd/src/branch/master/hooks/charmhelpers/contrib/openstack/ha/utils.py#L371).

File: src/config.yaml
Comment (sudeephb) [reply src/config.yaml:3]: It's not ready yet, The snap PR hasn't been merged

File: src/reactive/prometheus-apt-exporter.py
Comment (sudeephb) [reply src/reactive/prometheus-apt-exporter.py:31]: Turns out, if we drop this, the charm keeps re-installing itself forever

File: src/reactive/prometheus-apt-exporter.py
Comment (mkalcok) [inline src/reactive/prometheus-apt-exporter.py]: I think it's better to call this in `set_http_port` function. Along with `hookenv.close_port()` on the old port value. Otherwise the open ports will just keep appending.

File: src/reactive/prometheus-apt-exporter.py
Comment (mkalcok) [inline src/reactive/prometheus-apt-exporter.py:140]: This check should probably get reconfigured on port config change.

**Review by mkalcok — CHANGES_REQUESTED** [review]
Last two requests inline.

**Review by mkalcok — APPROVED** [review]
LGTM. Thanks. Please open issues for the two remaining unresolved conversation and then resolve them.

File: CONTRIBUTING.md
Comment (esunar) [inline CONTRIBUTING.md]: I think we should reference make

File: CONTRIBUTING.md
Comment (esunar) [inline CONTRIBUTING.md]: this is also wrong however we should reference make

File: CONTRIBUTING.md
Comment (esunar) [inline CONTRIBUTING.md]: we should use `make build`

**Review by esunar — CHANGES_REQUESTED** [review]
We should include `.gitignore`
remove `src/tests/unit/__pycache__/test_charm.cpython-310-pytest-7.2.0.pyc`

File: .gitignore
Comment (rgildein) [inline .gitignore:1]: When I saw it, I realized that we need to have template also for `.gitignore`.  [MP](https://code.launchpad.net/~rgildein/bootstack-charms-spec/+git/bootstack-charms-spec/+merge/434795)

This not blocking this PR to be merged.

**Review by rgildein — APPROVED** [review]
LGTM

**Review by esunar — APPROVED** [review]

**Pull request merged**
